### ISSUE TITLE
Performance Issue compared to Numpy[BUG]

### ISSUE THREAD

**Issue by NikhilKothari** [rapidsai/cudf#8313]
**Response time more than NumPy**
Indexing the cuDF DataFrame or cupy array takes almost double the time of that in numpy on Tesla P100 with CUDA 11.0

Example usage -

cuDF.loc[].reset_index(drop=True)
cuDF["column"][35:] = 0
cupy.array()[[0,5,9]]

All three above are way slower than numpy even though numpy runs on cpu.
Any reason in particular or am I using it wrong?

Thanks

**Comment by brandon-b-miller** [comment]
Hi @NikhilKothari, would you be able to share the code you are using to time/measure the performance? The best thing would a working example, maybe with some fake data - from there I can probably help shed some light on what you're seeing.

**Comment by NikhilKothari** [comment]
Sure, @brandon-b-miller. Below, is a code template that shows exactly what I am trying to do. I've added my Numpy version as well as the cuDF version of the solution for comparison. I'm creating some dummy data first and then feeding it to both versions.

```
import numpy as np
import random
import torch
import cudf

# corpus we need for the numpy version
corpus_for_numpy = np.array([['BRANDON', 'NIKHIL'] * 150000,
                             ['brandon-b-miller', 'NikhilKothari'] * 150000,
                             ['1', '0'] * 150000])

# the same corpus as above for cudf version
corpus_for_cudf = cudf.DataFrame({"person_info": ['BRANDON', 'NIKHIL'] * 150000, 
                                  "git_user_name": ['brandon-b-miller', 'NikhilKothari'] * 150000, 
                                  "score": [1, 0] * 150000})

# indices coming from pytorch on GPU
# need to index the corpus with these
idx = torch.tensor(random.sample(range(0, 299999), 200), device='cuda')

query = 'BRAN'


# numpy version of the solution
def numpy_version(corpus, idx, query):
  cut = corpus[:, idx.tolist()]
  cut[2, 35:] = "0.0"

  match = np.char.find(cut[0], query)
  score = (
      (match >= 0) * 10
      + cut[2].astype(float)
  ) / 11

  result = []

  for i in range(200):
      res = {
          "person_info": cut[0][i],
          "git_user_name": cut[1][i],
          "score": score[i],
      }

      result.append(res)

  return result

# cudf version of the same solution
def cudf_version(corpus, idx, query):
  cut = corpus.iloc[idx]
  cut["score"].iloc[35:] = 0
  cut["score"] = (cut["person_info"].str.contains(query) * 10 + cut["score"]) / 11
  # return cut.to_json(orient="records") # ignoring since this happens on cpu and takes even longer
  return cut
```

I'm calculating the response time in 3 ways. I've added them below with their values as comments.

```
%timeit numpy_version(corpus_for_numpy, idx, query) # 1000 loops, best of 5: 744 µs per loop
%timeit cudf_version(corpus_for_cudf, idx, query) # 100 loops, best of 5: 5.23 ms per loop
```

```
import time
now = time.time()
numpy_version(corpus_for_numpy, idx, query)
print(time.time() - now) # 0.002432107925415039

now = time.time()
cudf_version(corpus_for_cudf, idx, query)
print(time.time() - now) # 0.012674570083618164
```

```
from cupyx.time import repeat

# numpy_version       :    CPU:  745.985 us   +/-125.503 (min:  606.178 / max: 1721.858) us     GPU-0:  760.797 us   +/-127.248 (min:  617.024 / max: 1743.136) us
print(repeat(numpy_version, (corpus_for_numpy, idx, query), n_repeat=1000)) 

# cudf_version        :    CPU: 5373.742 us   +/-351.156 (min: 4842.586 / max: 8526.965) us     GPU-0: 5389.044 us   +/-351.938 (min: 4855.136 / max: 8560.512) us
print(repeat(cudf_version, (corpus_for_cudf, idx, query), n_repeat=1000))
```

I also wanted to understand if cuDF is built for real-time use cases in production. My Numpy solution runs on CPU and though it is fast, it chokes at 6k rpm while load testing. I want to perform the same operations on GPU and scale. Most of the GPU alternatives to Numpy don't work with strings. cuDF seemed to be an exception.

Any help, cuDF wise or otherwise would really be nice. I've been stuck on this since weeks now.

Thanks!

**Comment by brandon-b-miller** [comment]
So I played around a little with this today and while I am not sure I follow exactly what you are doing, I believe what you're seeing is basically a result of working with a rather small dataset. It appears early on in the workflow a subset of data is taken that is 200 elements in total and then performance is measured using that subset. 

Here are my results from comparing the runtimes while varying the size of the torch tensor across a range of values:

https://gist.github.com/brandon-b-miller/71ce300856666245103c785e236d4c1b

When using cuDF, or really any GPU based library, you'll generally see a higher speedup for larger datasets, whereas for small datasets, the CPU might be more performant.

**Comment by brandon-b-miller** [comment]
@NikhilKothari does the above help with your issue? If not I can dig a little deeper, if yes I'd like to resolve this ticket.

**Comment by NikhilKothari** [comment]
Definitely, @brandon-b-miller. Thank you for this clarification, it helps immensely. It would be fine to resolve the ticket.

I have one question though. Since cuDF uses GPU, would it be safe to assume that it would scale better on small datasets but high load though it is slower than Numpy? (Numpy's response time increases multifold as the load increases)

**Comment by brandon-b-miller** [comment]
Glad I could help :) Very generally one can expect the perf advantage vs the equivalent numpy op to grow as the data grows, although this is highly workload dependent. I recommend carefully benchmarking each workflow independently to figure out where the bottlenecks really are - not every workflow is suitable for GPU acceleration and in those cases the juice might not be worth the squeeze.
